REGRESSION(277450@main): OOB array read with SVG animation where keyPoints = 0.

mscottapple opened the pull request:
#### 53e67f679530f17b29fa09c1ca2af78cb1b0e0e4
<pre>
REGRESSION(277450@main): OOB array read with SVG animation where keyPoints = 0.
<a href="https://bugs.webkit.org/show_bug.cgi?id=272929">https://bugs.webkit.org/show_bug.cgi?id=272929</a>
<a href="https://rdar.apple.com/126636733">rdar://126636733</a>

Reviewed by Said Abou-Hallawa.

This change makes a couple additional, similar changes to the original changes
to better track the SVG spec. (See the original bug for more information.)

* LayoutTests/svg/animations/animate-zero-keyPoints-should-not-crash-expected.txt: Added.
* LayoutTests/svg/animations/animate-zero-keyPoints-should-not-crash.html: Added.
* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::keyTimes const):
(WebCore::SVGAnimationElement::startedActiveInterval):

Canonical link: <a href="https://commits.webkit.org/278212@main">https://commits.webkit.org/278212@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1a7f24f7858af91980c66c3e844390c392b9bec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49785 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29074 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/1994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53031 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52089 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35103 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/31 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40627 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51885 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26622 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21743 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24073 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30 "Found 1 new test failure: http/tests/media/hls/track-webvtt-multitracks.html (failure)") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8158 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/42 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54612 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24879 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/33 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48013 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26139 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42979 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47040 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10937 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26994 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25867 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGAnimationElement.cpp:
```diff
@@ -428,10 +428,8 @@ float SVGAnimationElement::calculatePercentFromKeyPoints(float percent) const
-    ASSERT(keyTimes.size() > 1);
-    ASSERT(m_keyPoints.size() == keyTimes.size());
+    if (m_keyPoints.isEmpty() || (calcMode() == CalcMode::Paced && animationMode() != AnimationMode::Path) || keyTimes.size() <= 1 || m_keyPoints.size() != keyTimes.size())
+        return percent;
```
Do we understand why we end up having any of these checks to be true?

According to the existing assertions these checks can't be true. Your change will silence these assertions forever. Are we sure we are not hiding an existing bug somewhere else?

mscottapple replied:
Fair point. I was trying to harden the function but for this one specific crash the issue is that `m_keyPoints.size()` and `keyTimes.size()` are < 2, which leads to the OOB. From the [commit message](https://github.com/WebKit/WebKit/pull/27204) of the change that introduced this issue,  this assertion is no longer guaranteed, at least for certain modes.

shallawa commented on LayoutTests/platform/gtk-wk2/svg/animations/animate-zero-keyPoints-should-not-crash-expected.txt:
```diff
@@ -0,0 +1,15 @@
+layer at (0,0) size 800x600
```
This is a crashing bug. We do not need to dump the layout tree in this case. All we need is to be able to print a text message saying "We did not crash". We should only have one text expectation file.

shallawa commented on LayoutTests/svg/animations/animate-zero-keyPoints-should-not-crash.html:
```diff
@@ -0,0 +1,3 @@
+<svg zoomAndPan="magnify">
```
It it is good to have a well-formed layout test. I would suggest using something like this:

```
<body>
    <svg>
        <text>
            Passes if it does not crash.
            <animateMotion path="m 1,0" keyPoints="0" keyTimes="0" restart="always"/>
        </text>
    </svg>
    <script>
        if (window.testRunner)
            testRunner.dumpAsText();
    </script>
</body>
```
There is no need to add the attribute `zoomAndPan="magnify"`. I think the bug can be reproduced without it.

mscottapple replied:
Oh, duh! I'll double-check that the `zoomAndPan` thing isn't needed and update the test.

shallawa commented on Source/WebCore/svg/SVGAnimationElement.cpp:
```diff
@@ -430,8 +430,9 @@ float SVGAnimationElement::calculatePercentFromKeyPoints(float percent) const
-    ASSERT(keyTimes.size() > 1);
     ASSERT(m_keyPoints.size() == keyTimes.size());
+    if (keyTimes.size() <= 1)
+        return percent;
```
I do not think this is the right approach to fix this bug. According to the [specs](https://svgwg.org/specs/animations/#AnimateMotionElement) and in the keyPoints section:

> If a list of ‘[keyPoints](https://svgwg.org/specs/animations/#KeyPointsAttribute)’ is specified, there must be exactly as many values in the ‘[keyPoints](https://svgwg.org/specs/animations/#KeyPointsAttribute)’ list as in the ‘[keyTimes](https://svgwg.org/specs/animations/#KeyTimesAttribute)’ list.

> If there are any errors in the ‘[keyPoints](https://svgwg.org/specs/animations/#KeyPointsAttribute)’ specification (bad values, too many or too few values), then the document is in error (see [Error processing](https://svgwg.org/svg2-draft/implnote.html#ErrorProcessing)).

The good news is `SVGAnimationElement::startedActiveInterval()` validates the values of the animation attributes and sets the `m_animationValid` accordingly. So all we need is to make sure `m_animationValid` to set to false if `keyPoints` and `keyTimes` don't match. If we do this then there is no need to change this function at all.

To do this I think you will need to change this line in `SVGAnimationElement::startedActiveInterval()`

```
m_animationValid = calcMode == CalcMode::Paced || !hasAttributeWithoutSynchronization(SVGNames::keyPointsAttr) || (keyTimes.size() > 1 && keyTimes.size() == m_keyPoints.size());
```

to be 

```
m_animationValid = !hasAttributeWithoutSynchronization(SVGNames::keyPointsAttr) || (keyTimes.size() > 1 && keyTimes.size() == m_keyPoints.size());
```

The reason for removing this clause is having `animationMode == AnimationMode::Path` makes `calcMode == CalcMode::Paced` to be ignored and we end up using the `keyPoints`. So we have to ignore `calcMode == CalcMode::Paced` when we check the validaty the attributes of `animationMode == AnimationMode::Path`.

mscottapple replied:
That does seem to work. PR updated.

shallawa commented on Source/WebCore/svg/SVGAnimationElement.cpp:
```diff
@@ -442,7 +442,6 @@ float SVGAnimationElement::calculatePercentFromKeyPoints(float percent) const
     if (calcMode() == CalcMode::Discrete)
         return fromKeyPoint;
     
-    ASSERT(index + 1 < keyTimes.size());
```
Why is this assertion removed? 

Looking at the logic in this function, I think it is only called when `keyTimes.size() > 1`. And `SVGAnimationElement::calculateKeyTimesIndex()` returns always value less than `keyTimes.size() - 1` if `keyTimes.size() > 1` and `calcMode() != CalcMode::Discrete`.

So this assertion should never fire.

mscottapple replied:
That was a mistake while backing out my prior changes.

mscottapple replied:
I have restored it, as well as fix the commit message to refer to the correct function for these new changes.

shallawa commented on Source/WebCore/svg/SVGAnimationElement.cpp:
```diff
@@ -393,7 +393,7 @@ static inline double solveEpsilon(double duration) { return 1 / (200 * duration)
 const Vector<float>& SVGAnimationElement::keyTimes() const
 {
-    return calcMode() == CalcMode::Paced ? m_keyTimesForPaced : m_keyTimesFromAttribute;
+    return (calcMode() == CalcMode::Paced && animationMode() != AnimationMode::Path) ? m_keyTimesForPaced : m_keyTimesFromAttribute;
```
Nit: There is no need for the parentheses. The operator precedence should work as expected:

1. The equality != and == 
2. The logical &&
3. The ternary conditional ? :

mscottapple committed:
REGRESSION(277450@main): OOB array read with SVG animation where keyPoints = 0.
https://bugs.webkit.org/show_bug.cgi?id=272929
rdar://126636733

Reviewed by Said Abou-Hallawa.

This change makes a couple additional, similar changes to the original changes
to better track the SVG spec. (See the original bug for more information.)

* LayoutTests/svg/animations/animate-zero-keyPoints-should-not-crash-expected.txt: Added.
* LayoutTests/svg/animations/animate-zero-keyPoints-should-not-crash.html: Added.
* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::SVGAnimationElement::keyTimes const):
(WebCore::SVGAnimationElement::startedActiveInterval):

Canonical link: https://commits.webkit.org/278212@main
